### DISCUSSION TITLE
DRY include with type alters query return type? Why? Workaround?

### DISCUSSION THREAD

**Discussion by hexpunk** [prisma/prisma#18284] [Q&A] [answered]
Here's a simplified version of a query I recently wrote.

```typescript
const book1 = await prisma.book.findUniqueOrThrow({
  where: {id: 1},
  include: {
    authors: true,
    publisher: true,
  },
});

/*
 * The type of `book1` is:
 * Book & {
 *   authors: Author[];
 *   publisher: Publisher;
 * }
 */
```

It's important to note that as I typed the `include` object in the call to `findUniqueOrThrow`, auto-complete/intellisense/whatever in VSCode was able to show me valid keys I could use since it inferred the type of `include`.

In my particular case, I found myself needing to make many queries using the same `include` block.  (The real `include` block was much more complicated than this simplified example.)  I made a variable to reuse for every query.

```typescript
const include = {
  authors: true,
  publisher: true,
};

const book2 = await prisma.book.findUniqueOrThrow({
  where: {id: 2},
  include,
});

const book3 = await prisma.book.findUniqueOrThrow({
  where: {id: 3},
  include,
});

// etc.

// Note that the types of `book2` and `book3` are identical to that of `book1` in the previous example.
```

This worked as expected, but I found myself missing the auto-complete feature as I typed out the `include` variable this time.  I understand why Typescript can't predict the valid key names and value types since `include` is inferred at this point.  To remedy this, I explicitly set the type of `include` to `Prisma.BookInclude`.

```typescript
const include: Prisma.BookInclude = {
  authors: true,
  publisher: true,
};

const book4 = await prisma.book.findUniqueOrThrow({
  where: {id: 4},
  include,
});

/*
 * The type of `book4` is:
 * Book & {
 *   authors?: Author[] | undefined;
 *   publisher?: Publisher | null | undefined;
 * }
 */
```

Auto-complete now works as expected as I fill out the value for `include`.  However, the return type of `findUniqueOrThrow` has changed in an undesirable way.  Loaded relationships are now possibly `undefined`, meaning I'll have to use the optional chaining operator (`?.`) or the non-null assertion operator (`!`) to safely access deep values.

My best guess as to why this is happening is that `BookInclude` is a wider type than what Typescript would infer `include`'s type to be.  If my options are to lose intellisense on my `include` object or lose a degree of type specificity on the return type of query functions, the choice is easy.  I'll forfeit intellisense on `include` every time.

Is there a way I can have both?  Some type I can give to `include` that won't make all my eagerly loaded relationships possibly `undefined`?

**Comment by nurul3101** [comment, marked as answer]
Hello @hexpunk 👋 

Instead of defining the type manually, can you try using Prisma.validator API to define the types?
Here's a [reference](https://www.prisma.io/docs/concepts/components/prisma-client/advanced-type-safety/operating-against-partial-structures-of-model-types#problem-using-variations-of-the-generated-model-type) on how to declare types with included relation.

**Reply by hexpunk** [reply]
Thanks, @nurul3101.  After reading [this page](https://www.prisma.io/docs/concepts/components/prisma-client/advanced-type-safety/prisma-validator) specifically, I believe what you're recommending is the correct answer.

However, I did notice that for any given object, VSCode's autocomplete would only offer suggestions for the first key I typed.  It wouldn't offer any suggestions for further keys.

```typescript
const include = Prisma.validator<Prisma.BookInclude>()({
  authors: true,
  publisher: true,
});

const book4 = await prisma.book.findUniqueOrThrow({
  where: {id: 4},
  include,
});
```

For example, in this case `book4` was of the expected type.  And when I started typing `authors`, autocomplete suggested the key.  But when I started typing `publisher`, autocomplete offered no keys at all.  This pattern was consistent if I added the `publisher` key first and the `authors` key second.  The first key was suggested and the second (third, fourth, etc) were treated as if VSCode was unaware of them until I typed out the entire key by hand.

That's not ideal, but I can't think of any reason Prisma would be at fault for this.  It must be a bug in either the Typescript language server or VSCode itself.
